### PR TITLE
feat: option to strictly follow Go autogenerated file convention

### LINKED ISSUE #4506: Issues in gqlgen resolvers are ignored
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

I'm using [gqlgen](https://gqlgen.com) to implement a GraphQL server. It autogenerates stubs for GraphQL endpoints, in `*.resolver.go` files. Even though the stubs are autogenerated, the files have non-autogenerated code.

`golangci-lint run` ignores issues in these files. Based on the verbose output, I think they're getting filtered out by the `autogenerated_exclude` processor.

Each resolver file has a comment like this:
```go
// This file will be automatically regenerated based on the schema, any resolver implementations
// will be copied through when generating and any unknown code will be moved to the end.
// Code generated by github.com/99designs/gqlgen version v0.17.27
```

I think that's triggering the rule [here](https://github.com/golangci/golangci-lint/blob/c6ac5a12b9db7eaecf38d8f60950e93741f4444a/pkg/result/processors/autogenerated_exclude.go#L77), even though the rest of the file isn't autogenerated.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.56.2 built with go1.22.0 from 58a724a on 2024-02-15T12:52:06Z
```

</details>

### Configuration

<details>

```yaml
run:
linters:
  enable:
    - gocritic
    - exhaustruct
linters-settings:
  exhaustruct:
    exclude:
      - '.+/graph\.Config$'
      - '.+/graphql\.FieldContext$'
      - '.+/graphql\.Response$'

```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.22.1 darwin/arm64
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/ben/Library/Caches/go-build'
GOENV='/Users/ben/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/ben/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/ben/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.22.1/libexec'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.22.1/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.22.1'
GCCGO='gccgo'
AR='ar'
CC='cc'
CXX='c++'
CGO_ENABLED='1'
GOMOD='/Users/ben/src/gqlgen-linting/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/vx/hb7qrv412lnbfgmqn9rpwc200000gn/T/go-build1658647304=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/ben/src/gqlgen-linting /Users/ben/src /Users/ben /Users /]
INFO [config_reader] Used config file .golangci.yaml
INFO [lintersdb] Active 8 linters: [errcheck exhaustruct gocritic gosimple govet ineffassign staticcheck unused]
INFO [loader] Go packages loading at mode 575 (compiled_files|exports_file|name|deps|files|imports|types_sizes) took 144.178708ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 2.938416ms
INFO [linters_context/goanalysis] analyzers took 3.379248363s with top 10 stages: buildir: 2.373889789s, inspect: 125.784248ms, fact_deprecated: 103.66313ms, fact_purity: 100.208455ms, ctrlflow: 96.994997ms, printf: 92.911998ms, nilness: 77.548118ms, SA5012: 77.15329ms, typedness: 68.571036ms, gocritic: 24.120708ms
INFO [runner] Issues before processing: 1, after processing: 0
INFO [runner] Processors filtering stat (out/in): cgo: 1/1, skip_files: 1/1, path_prettifier: 1/1, skip_dirs: 1/1, filename_unadjuster: 1/1, autogenerated_exclude: 0/1
INFO [runner] processing took 61.588µs with stages: autogenerated_exclude: 27.625µs, path_prettifier: 24.625µs, skip_dirs: 4.501µs, cgo: 625ns, nolint: 625ns, max_same_issues: 416ns, max_from_linter: 376ns, fixer: 375ns, severity-rules: 334ns, filename_unadjuster: 333ns, exclude-rules: 292ns, uniq_by_line: 292ns, skip_files: 209ns, exclude: 209ns, identifier_marker: 167ns, source_code: 167ns, path_shortener: 126ns, diff: 125ns, sort_results: 83ns, max_per_file_from_linter: 42ns, path_prefixer: 41ns
INFO [runner] linters took 1.274062792s with stages: goanalysis_metalinter: 1.273950375s
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 16 samples, avg is 212.5MB, max is 366.9MB
INFO Execution took 1.427133584s
```

</details>

<details>
<summary>Running `exhaustruct` directly:</summary>

```console
$ exhaustruct -e '.+/graph\.Config$' -e '.+/graphql\.FieldContext$' -e '.+/graphql\.Response$' ./...
/Users/ben/src/gqlgen-linting/graph/schema.resolvers.go:21:11: model.Todo is missing fields ID, Done, User
```
</details>

### A minimal reproducible example or link to a public repository

See https://github.com/bnavetta/gqlgen-linting

The code that should be reported is here: https://github.com/bnavetta/gqlgen-linting/blob/2f18e1ecca7cffa1030a12aa10d5af4e1ce49de4/graph/schema.resolvers.go#L21-L24

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

### PR DESCRIPTION
**Pull request by ldez**
The current autogenerated file detection is very lax and it's expected (#48, #72).

But in some cases, this can be a problem.

Then I added an option to strictly follow Go autogenerated file convention.

> To convey to humans and machine tools that code is generated, generated source should have a line that matches the following regular expression (in Go syntax):
>
> `^// Code generated .* DO NOT EDIT\.$`
>
> This line must appear before the first non-comment, non-blank text in the file. 
> https://go.dev/s/generatedcode

The strict mode can be extended with simple `run.skip-dirs` or `run.skip-files`.

One commit, one change.

Fixes #4506

### CODE REVIEW

File: jsonschema/golangci.next.jsonschema.json [hunk 1]
```diff
@@ -3342,6 +3342,11 @@
           "type": "boolean",
           "default": false
         },
+        "exclude-autogenerated-strict": {
+          "description": "To follow strict Go autogenerated file convention",
+          "type": "boolean",
+          "default": false
+        },
```
Comment (bombsimon) [inline jsonschema/golangci.next.jsonschema.json:3349]: So nice to see this in the same PR (even though manual work for now). 🎈

File: pkg/result/processors/autogenerated_exclude.go [hunk 2]
```diff
@@ -6,32 +6,43 @@ import (
...
-var autogenDebugf = logutils.Debug(logutils.DebugKeyAutogenExclude)
+const (
+	genCodeGenerated = "code generated"
+	genDoNotEdit     = "do not edit"
+	genAutoFile      = "autogenerated file" // easyjson
+)
+
+var _ Processor = &AutogeneratedExclude{}
 
 type ageFileSummary struct {
```
Comment (bombsimon) [inline pkg/result/processors/autogenerated_exclude.go:24]: What does `age` refer to? 🤔

**Review by bombsimon — APPROVED** [review]
LGTM!

Would you like to deprecate the old variant at some point? It feels like supporting anything that is not the Go convention is a bad idea and if you want to do that it would make more sense to use `skip-dirs` or `skip-files` but I also see the convenience keeping it and having the strict version as opt-in.

**Comment by ldez** [comment]
> Would you like to deprecate the old variant at some point?

Maybe just change the default, because I think some people relies on that.

> it would make more sense to use `skip-dirs` or `skip-files`

I think `skip-dirs` or `skip-files` should be deprecated and replaced by something in the `issues` section with a relevant name.
Because `skip-xxx` are skipping nothing :smile:

File: pkg/result/processors/autogenerated_exclude.go (hunk 2 above)
Comment (ldez) [reply pkg/result/processors/autogenerated_exclude.go]: I renamed the structure.

File: .golangci.next.reference.yml [hunk 3]
```diff
@@ -2814,6 +2814,12 @@ issues:
   # Default: false
   exclude-case-sensitive: false
 
+  # To follow strict Go autogenerated file convention.
+  # https://go.dev/s/generatedcode
+  # By default a lax pattern is applied.
+  # Default: false
+  exclude-autogenerated-strict: true
```
Comment (alexandear) [inline .golangci.next.reference.yml:2821]: Should we add a new option that allows linting of the generated files?
```
# If exclude generated files.
# Default: true
exclude-generated: false
```
Of course, this is a separate feature.

File: pkg/result/processors/autogenerated_exclude.go [hunk 4]
```diff
@@ -56,66 +63,96 @@ func (p *AutogeneratedExclude) shouldPassIssue(issue *result.Issue) (bool, error
...
+
 	return false
 }
 
-func (p *AutogeneratedExclude) getOrCreateFileSummary(issue *result.Issue) (*ageFileSummary, error) {
-	fs := p.fileSummaryCache[issue.FilePath()]
-	if fs != nil {
-		return fs, nil
+// Based on https://go.dev/s/generatedcode
+// > This line must appear before the first non-comment, non-blank text in the file.
```
Comment (alexandear) [inline pkg/result/processors/autogenerated_exclude.go:126]: I propose extending the comment:
```go
// isGeneratedFileStrict returns true if the source file has a line that matches the regular expression:
//
//	^// Code generated .* DO NOT EDIT\.$
//
// This line must appear before the first non-comment, non-blank text in the file.
// Based on https://go.dev/s/generatedcode.
```

File: pkg/result/processors/autogenerated_exclude.go [hunk 5]
```diff
@@ -56,66 +63,96 @@ func (p *AutogeneratedExclude) shouldPassIssue(issue *result.Issue) (bool, error
...
-// match more generated code. See #48 and #72.
-func isGeneratedFileByComment(doc string) bool {
-	const (
-		genCodeGenerated = "code generated"
-		genDoNotEdit     = "do not edit"
-		genAutoFile      = "autogenerated file" // easyjson
-	)
-
+// isGeneratedFileLax reports whether the source file is generated code.
+// Using a bit laxer rules than https://go.dev/s/generatedcode to match more generated code.
```
Comment (alexandear) [inline pkg/result/processors/autogenerated_exclude.go:105]: I think we can refer to the `isGeneratedFileStrict` function here:
```suggestion
// The function uses a bit laxer rules than isGeneratedFileStrict to match more generated code.
```

File: pkg/result/processors/autogenerated_exclude.go [hunk 6]
```diff
@@ -56,66 +63,96 @@ func (p *AutogeneratedExclude) shouldPassIssue(issue *result.Issue) (bool, error
...
+	if p.strict {
+		var err error
+		fs.generated, err = p.isGeneratedFileStrict(issue.FilePath())
+		if err != nil {
+			return false, fmt.Errorf("failed to get doc of file %s: %w", issue.FilePath(), err)
+		}
+	} else {
+		doc, err := getComments(issue.FilePath())
+		if err != nil {
+			return false, fmt.Errorf("failed to get doc of file %s: %w", issue.FilePath(), err)
```
Comment (alexandear) [inline pkg/result/processors/autogenerated_exclude.go:92]: To simplify future debugging, we should use different errors for strict and non-strict paths.
Currently, if either `isGeneratedFileStrict` or `getComments` fails, the error returned is `fmt.Errorf("failed to get doc of file %s: %w", issue.FilePath(), err)`.

File: pkg/result/processors/autogenerated_exclude_test.go [hunk 7]
```diff
@@ -2,74 +2,107 @@ package processors
...
+	p := NewAutogeneratedExclude(true)
+
+	testCases := []struct {
+		desc     string
+		filepath string
+		assert   assert.BoolAssertionFunc
+	}{
+		{
+			desc:     "",
+			filepath: filepath.FromSlash("./testdata/autogen_go_strict.go"),
```
Comment (alexandear) [inline pkg/result/processors/autogenerated_exclude_test.go:82]: Can be shorter?
```suggestion
			filepath: filepath.FromSlash("testdata/autogen_go_strict.go"),
```

File: pkg/result/processors/autogenerated_exclude.go [hunk 8]
```diff
@@ -40,11 +51,7 @@ func (p *AutogeneratedExclude) Process(issues []result.Issue) ([]result.Issue, e
...
 }
 
-func isSpecialAutogeneratedFile(filePath string) bool {
-	fileName := filepath.Base(filePath)
-	// fake files or generation definitions to which //line points to for generated files
-	return filepath.Ext(fileName) != ".go"
-}
+func (p *AutogeneratedExclude) Finish() {}
 
 func (p *AutogeneratedExclude) shouldPassIssue(issue *result.Issue) (bool, error) {
```
Comment (alexandear) [inline pkg/result/processors/autogenerated_exclude.go:56]: Perhaps we should cover this function with some tests due to the new `strict` logic.

Or even better, we should add tests for the exported `AutogeneratedExclude.Process` function.

**Pull request merged**